ocaml/xapi/xapi_vtpm.ml

```diff
@@ -20,20 +20,16 @@ let assert_no_vtpm_associated ~__context vm =
       let amount = List.length vtpms |> Int.to_string in
       raise Api_errors.(Server_error (vtpm_max_amount_reached, [amount]))
 
-let introduce ~__context ~uuid ~vM ~backend ~profile ~contents =
-  let ref = Ref.make () in
-  Db.VTPM.create ~__context ~ref ~uuid ~vM ~backend ~profile ~contents ;
-  ref
-
 let create ~__context ~vM =
   assert_no_vtpm_associated ~__context vM ;
   Xapi_vm_lifecycle.assert_initial_power_state_is ~__context ~self:vM
     ~expected:`Halted ;
+  let ref = Ref.make () in
   let uuid = Uuid.(to_string (make ())) in
-  let profile = Db.VM.get_default_vtpm_profile ~__context ~self:vM in
   let backend = `xapi in
   let contents = Xapi_secret.create ~__context ~value:"" ~other_config:[] in
-  let ref = introduce ~__context ~uuid ~vM ~backend ~profile ~contents in
+  Db.VTPM.create ~__context ~ref ~uuid ~vM ~backend ~unique:false
```

Comment:
We should add unique and protected as parameters to create, defaulting to false if absent.

Reply:
While I agree on the unique field, I do not think it makes sense to give control to users of protected. It's a property of how the system handles the vtpm contents and should be read-only.

For the unique field I expected to defer this until it's properly implemented. I can change the constructor and ignore the value of the field for the time being.

Reply:
They are readonly after the object is created, but how else would the user chose whether they want a protected vTPM or not if not at create time?

Reply:
I would say that the whole point of TPMs is to have some secrets protected, so I don't see the point in making users choose. I see it as a way to report the system property.

(I'm working on exposing the unique field in the create)

Reply:
I've added the unique parameter, there's no easy way to have it as an option: having option types in `create` functions is disallowed by the api generator, and because a unit parameter needs to be added to remove ambiguity cause by partial application.

I believe this is not such a bad compromise:
- for backwards compatibility we have the platform metadata in templates, which will always create non-unique vtpms, this is fine because this is wanted only to be able to install / boot windows 11 guests.
- New clients will need an updated SDK to create VTPMs anyway, this will make the option obvious to them and can make choices for their users.

Reply:
Depending on how we implement protection (e.g. per-host or per-pool encryption key) then a VM exported from a pool will only be reimportable on the same pool (unless you provide the encryption keys in some way, but those should be non-exportable...). 
This is fine if you've actually stored secrets in the vTPM, however if you only want to share a Windows11 VM image between multiple pools, which has a vTPM (not for security purposes, but purely to satisfy Win11 requirements), then having it marked as protected will make that quite difficult (in particular none of the existing tooling/applications will work with it because they wouldn't know a key has to be supplied or how to supply it).
In that case you might want to create the VM with unprotected vTPM as a functionality/security tradeoff.

*or* we should provide a way to export a "wiped"/empty vTPM/vTPM with just public certs, in case a protected vTPM is being exported, acknowledging that this may lead to data loss if the VM has actually stored some secrets or data inside the vTPM.